### PR TITLE
Expansions

### PR DESCRIPTION
**Pull request by amaanq**
@ObserverOfTime

### CODE REVIEW

**Comment by amaanq** [comment]
`${!x//c/x}` is tough

**Comment by ObserverOfTime** [comment]
> `${!x//c/x}` is tough

I don't think that's valid...

**Comment by amaanq** [comment]
it's in the corpus

**Comment by ObserverOfTime** [comment]
Maybe you could make expansions start with `optional('!')` with the exception of `${#foo}`, `${!foo*}` & `${!foo@}`.

**Comment by amaanq** [comment]
${!##} is doing my head in

**Comment by amaanq** [comment]
after a few painful hours I think I have a complete solution, would appreciate you trying it out @ObserverOfTime

File: test/corpus/literals.txt [hunk 1]
```diff
@@ -454,6 +432,131 @@ cat ${BAR} ${ABC=def} ${GHI:?jkl}
...
+  (command (command_name (expansion)))
+  (command (command_name (expansion)))
+  (command (command_name (expansion (special_variable_name) (regex)))))
+
+================================================================================
+Variable Expansions: Weird Cases
+================================================================================
+
+${completions[*]}
+${=1}
```
Comment (ObserverOfTime) [inline test/corpus/literals.txt:509]: `bash: ${=1}:: bad substitution`

Comment (amaanq) [reply test/corpus/literals.txt:509]: It's valid in here, let me find the example file that had it

```bash
for c in ${=1}; do
	echo c
done
```

Comment (amaanq) [reply test/corpus/literals.txt:509]: http://www.oilshell.org/git-branch/master/21552818/wild.wwz/linux-4.8.7/tools/perf/perf-completion.sh.txt

Comment (ObserverOfTime) [reply test/corpus/literals.txt:509]: Interesting. I can't find it anywhere in the bash manual. :thinking:

Comment (amaanq) [reply test/corpus/literals.txt:509]: maybe it's in an older version

**Comment by ObserverOfTime** [comment]
Expansions appear to be fine now but there are a few arithmetic issues left


- Operators in C-style binary expressions are not `operator` fields.
- The unary expression operators `"-" "+" "~"` cannot be captured.
  - Maybe unary and postfix operators should also be fields.
- `test_operator` seems to have higher precedence over negative numbers.

```bash
echo $((- f))
echo $((-f))
```

<details>
<summary>Parse tree</summary>

```
0:0  - 2:0     program
0:0  - 0:13      command
0:0  - 0:4         name: command_name
0:0  - 0:4           word `echo`
0:5  - 0:13        argument: arithmetic_expansion
0:5  - 0:8           "$(("
0:8  - 0:11          unary_expression
0:10 - 0:11            variable_name `f`
0:11 - 0:13          "))"
0:13 - 1:0       "\n"
1:0  - 1:12      command
1:0  - 1:4         name: command_name
1:0  - 1:4           word `echo`
1:5  - 1:12        argument: arithmetic_expansion
1:5  - 1:8           "$(("
1:8  - 1:10          test_operator `-f`
1:10 - 1:12          "))"
1:12 - 2:0       "\n"
```

</details>

**Comment by amaanq** [comment]
are test operators allowed in arithmetic expansions?

**Comment by ObserverOfTime** [comment]
Not as far as I can tell.

```bash
$  echo $((0 -eq 0))
bash: 0 -eq 0: syntax error in expression (error token is "0")
```

**Comment by amaanq** [comment]
all operators are fielded & should be exposed (incl. unary ops)

**Comment by amaanq** [comment]
Hmm interesting, token(prec(choice())) prevents querying the operators, but not choice(token(prec())) (no need for aliasing)

**Comment by ObserverOfTime** [comment]
I still can't seem to capture unary operators.

**Comment by amaanq** [comment]
got a query code? this works for me on this branch:

```scheme
(unary_expression "+" @op)
```

**Comment by ObserverOfTime** [comment]
Does it work with `x=$((+y))` ?

**Comment by amaanq** [comment]
damn, the query compiles but doesn't seem to execute properly, I'll bring back the aliases

**Comment by ObserverOfTime** [comment]
I will once again advocate for a `prec.lexical` function that is intuitive and works as expected.

**Comment by amaanq** [comment]
Nope, this was my fault. I forgot I had _arithmetic_unary_expression too

**Comment by amaanq** [comment]
try now @ObserverOfTime

**Comment by ObserverOfTime** [comment]
No field. :^)

**Comment by amaanq** [comment]
would you consider ternaries candidates to be fielded as operator?

**Comment by ObserverOfTime** [comment]
> would you consider ternaries as candidates to be fielded as operator?

It's fine either way.

**Comment by amaanq** [comment]
Alright, I don't really care to - so this should be good as is hopefully

**Comment by ObserverOfTime** [comment]
Yeah, finally bash is usable.

**Comment by amaanq** [comment]
Yeah it was pretty "eh" the few times I used it in the past :^) (and arithmetic expansions broke everything)

But writing the parser this language definitely sucks

**Comment by ObserverOfTime** [comment]
> But writing the parser this language definitely sucks

What do you mean this language sucks? :anger:

**Comment by amaanq** [comment]
pfft...it won't suck all that much compared to zsh :grimacing: :nauseated_face:

**Comment by ObserverOfTime** [comment]
Or powershell.

**Comment by amaanq** [comment]
as if I'd want to touch Windows :grin:

**Comment by verhovsky** [comment]
@amaanq powershell actually runs on Linux and macOS, it's even open source under the MIT license.

https://formulae.brew.sh/cask/powershell

https://github.com/PowerShell/PowerShell

**Comment by amaanq** [comment]
wow, I never knew, that's cool but I wonder if people actually use it (except for cross-platform testing)

**Pull request merged**